runner/webserver.ts

```diff
@@ -75,119 +58,129 @@ export function webserver(wct: Context): void {
-        response.set(DEFAULT_HEADERS);
-        send(request, file).pipe(response);
-      });
+    // Serve up project & dependencies via polyserve
+    const polyserveResult = await startServers({
+      root: options.root,
+      headers: DEFAULT_HEADERS,
+      packageName,
+      additionalRoutes: additionalRoutes,
     });
+    let servers: Array<MainlineServer|VariantServer>;
+    if (polyserveResult.kind === 'mainline') {
+      servers = [polyserveResult];
+      wsOptions.port = polyserveResult.server.address().port;
+    } else if (polyserveResult.kind === 'MultipleServers') {
+      servers = [polyserveResult.mainline];
+      servers = servers.concat(polyserveResult.variants);
+      wsOptions.port = polyserveResult.mainline.server.address().port;
+    } else {
+      const never: never = polyserveResult;
+      throw new Error(
+          `Internal error: Got unknown response from polyserve.startServers: ${never}`);
+    }
 
-    // The generated web runner, if present.
-    if (wsOptions.webRunnerContent) {
-      app.get(wsOptions.webRunnerPath, function(request, response) {
-        response.set(DEFAULT_HEADERS);
-        response.send(wsOptions.webRunnerContent);
+    const onDestroyHandlers: Array<() => Promise<void>> = [];
+    for (const server of servers) {
+      const destroyableServer = server.server as any;
+      serverDestroy(destroyableServer);
+      onDestroyHandlers.push(() => {
+        destroyableServer.destroy();
+        return new Promise<void>(
+            (resolve) => server.server.on('close', () => resolve()));
       });
     }
 
-    // At this point, we allow other plugins to hook and configure the
-    // webserver as they please.
-    await wct.emitHook('prepare:webserver', app);
-
-    // Serve up all the static assets.
-    app.use(serveWaterfall(wsOptions.pathMappings, {
-      root: options.root,
-      headers: DEFAULT_HEADERS,
-      log: wct.emit.bind(wct, 'log:debug'),
-    }));
+    wct._httpServers = servers.map(s => s.server);
 
-    app.use('/httpbin', httpbin.httpbin);
+    // At this point, we allow other plugins to hook and configure the
+    // webservers as they please.
+    for (const server of servers) {
+      await wct.emitHook('prepare:webserver', server.app);
+    }
 
-    app.get('/favicon.ico', function(request, response) {
-      response.end();
+    options.webserver._servers = servers.map(s => {
+      return {
+        url: `http://localhost:${s.server.address().port}${pathToGeneratedIndex
+                                                        }`,
+        variant: s.kind === 'mainline' ? '' : s.variantName
+      };
     });
 
-    app.use(function(request, response, next) {
-      wct.emit('log:warn', '404', chalk.magenta(request.method), request.url);
-      next();
-    });
+    // TODO(rictic): re-enable this stuff. need to either move this code into
+    //     polyserve or let the polyserve API expose this stuff.
+    // app.use('/httpbin', httpbin.httpbin);
```

Comment:
It isn't. httpbin is a Router, not a normal request handler, so it does its own sub-path matching. I tried a few things to get this to work with additionalRoutes and nothing seemed to work.

Filed as https://github.com/PolymerLabs/polyserve/issues/138